src/controller/java/BdxOTASender.cpp

```diff
@@ -0,0 +1,411 @@
+                        static_cast<jlong>(mNodeId));
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+
+    ResetState();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BdxOTASender::OnBlockQuery(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturnError(mFabricIndex != kUndefinedFabricIndex, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mNodeId != kUndefinedNodeId, CHIP_ERROR_INCORRECT_STATE);
+
+    uint16_t blockSize  = mTransfer.GetTransferBlockSize();
+    uint32_t blockIndex = mTransfer.GetNextBlockNum();
+
+    uint64_t bytesToSkip = 0;
+    if (event.EventType == TransferSession::OutputEventType::kQueryWithSkipReceived)
+    {
+        bytesToSkip = event.bytesToSkip.BytesToSkip;
+    }
+
+    // uint64_t transferGeneration = mTransferGeneration;
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    jmethodID handleBDXQueryMethod;
+    CHIP_ERROR err = JniReferences::GetInstance().FindMethod(
+        env, mOtaDelegate, "handleBDXQuery", "(JIJJ)Lchip/devicecontroller/OTAProviderDelegate$BDXData;", &handleBDXQueryMethod);
+    VerifyOrReturnError(err == CHIP_NO_ERROR, err, ChipLogError(Controller, "Could not find handleBDXQuery method"));
+
+    jobject bdxData =
+        env->CallObjectMethod(mOtaDelegate, handleBDXQueryMethod, static_cast<jlong>(mNodeId), static_cast<jint>(blockSize),
+                              static_cast<jlong>(blockIndex), static_cast<jlong>(bytesToSkip));
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+
+    if (bdxData == nullptr)
+    {
+        LogErrorOnFailure(mTransfer.AbortTransfer(bdx::StatusCode::kUnknown));
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    jmethodID getDataMethod;
+    err = JniReferences::GetInstance().FindMethod(env, bdxData, "getData", "()[B", &getDataMethod);
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+
+    jmethodID isEOFMethod;
+    err = JniReferences::GetInstance().FindMethod(env, bdxData, "isEOF", "()Z", &isEOFMethod);
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+    jbyteArray jData = (jbyteArray) env->CallObjectMethod(bdxData, getDataMethod);
+    jboolean jIsEOF  = env->CallBooleanMethod(bdxData, isEOFMethod);
+
+    JniByteArray data(env, jData);
```

Comment:
please stop using this way to manage the lifetime for localRef, which is leaking.

Reply:
This class can manage reference lifetime. Is there something I don't know?